### ISSUE TITLE
virtualNode hook?

### ISSUE THREAD

**Issue by codec-abc** [fable-compiler/fable-arch#67]
Hi,

I may have missed it, but is there a way to call some code when a virtual node is inserted into the real DOM? In my case, I have created a simple div that I want to turn into a IDE text-editor alike using [CodeMirror](https://codemirror.net/). For that I need to call a function of the library once my div has been inserted into the DOM but cannot find a proper way to do that.

**Comment by MangelMaxime** [comment]
Hello @codec-abc ,

VirtualDom is just mapping over the DOM so it's seems like you need to work with [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).

You could also create an action `checkDiv` which will test if the div is present in the DOM and init the CodeMirror otherwiser this action could call itself 100ms later and check again for example.

Another possibilities depending on your app, would be to have the Fable-Arch app and the Code Mirror app work side by side in the application. And if you need to communicate with one or the other use the producer/subscriber.

**Comment by codec-abc** [comment]
Thanks for the kind answer. Before opening the issue I read about [hooks in virtual-dom](https://github.com/Matt-Esch/virtual-dom/blob/master/docs/hooks.md) and wondered if there was a similar concept built-in in fable-arch. From your answer I guess that the answer is no (please correct me if it is otherwise)

About the solutions, I use the 2nd for now (the one with the timer) but the 1st seems more robust even if I am not sure about the performance implication to watch almost the entire DOM.

**Comment by MangelMaxime** [comment]
For the first solution, you should be able to watch only part of the application.
Because, normally you should have this order of operation:

Update function (which add actions to the stack) > Render the view > Process the next action in the stack

So if you attach the Observer inside the actions added to the stack you should be fine.

What I am saying need more test to be su

**Comment by MangelMaxime** [comment]
I just read again the doc about the hook.

I think this is possible to use `Hook` function. I will try to write a sample about it tomorrow.

**Comment by MangelMaxime** [comment]
Ok, just a quick update.
This is definitely possible, I am now working on a nice way to write the `Hook` function :)

**Comment by codec-abc** [comment]
To be honest the solution you gave me with the MutationObserver (which I didn't know because I am a html/js/css noob :) ) works great and the project for which I use fable-arch is nothing serious. So don't rush yourself just because of me opening an issue. That being said, thanks again for your support.

**Comment by MangelMaxime** [comment]
Oh no problem,
glad that I can help and it's good to no work on the docs site for once ^^

**Comment by MangelMaxime** [comment]
@mastoj I have one question about the hook implemetion.

I have it implemented on my local version of Fable-Arch.

Here is the helper proposition:

```fsharp
    type HookHelper =
        // Hook
        [<Emit("(function() { var Hook = function() {}; Hook.prototype.hook = $0; return new Hook(); })()")>]
        static member CreateHook(hook: JsFunc2<HTMLElement, string, unit>) : HookValue = jsNative
        [<Emit("(function() { var Hook = function() {}; Hook.prototype.hook = $0; return new Hook(); })()")>]
        static member CreateHook(hook: JsFunc3<HTMLElement, string, HTMLElement, unit>) : HookValue = jsNative
        // Unhook
        [<Emit("(function() { var Hook = function() {}; Hook.prototype.unhook = $0; return new Hook(); })()")>]
        static member CreateUnhook(unhook: JsFunc2<HTMLElement, string, unit>) : HookValue = jsNative
        [<Emit("(function() { var Hook = function() {}; Hook.prototype.unhook = $0; return new Hook(); })()")>]
        static member CreateUnhook(unhook: JsFunc3<HTMLElement, string, HTMLElement, unit>) : HookValue = jsNative
        // Hook & Unhook
        [<Emit("(function() { var Hook = function() {}; Hook.prototype.hook = $0; Hook.prototype.unhook = $1; return new Hook(); })()")>]
        static member CreateTwoWayHook(hook: JsFunc3<HTMLElement, string, HTMLElement, unit>, unhook: JsFunc3<HTMLElement, string, HTMLElement, unit>) : HookValue = jsNative
```

I propose an Helper using the type just to allow declaration of function with 2 or 3 arguments has the third is optional.
This is also possible to implement functions, but then we would always need to declare the 3 arguments I think.

Here is what a hook declaration could look in the view function

```fsharp
div 
  [ hook 
      "my-hook"
      (HookHelper.CreateHook(fun node propName -> console.log "Hello from the hook"))
    hook 
      "my-hook2"
      (HookHelper.CreateUnhook(fun node propName oldNode -> console.log "Hello from the unhook"))
    hook 
      "my-hook-two-way"
      (HookHelper.CreateTwoWayHook( 
        (fun node propName oldNode -> console.log "Hello from the hook two way"),
        (fun node propName oldNode -> console.log "Hello from the unhook two way")
      ))
  ]  
  [ text "Maxime" ] 
```

It can look a lot of code but you would never (I think) implement all this hook on one element :)

I have two questions:
1. Are you ok with the current proposition ? (Functions name, usage of type etc.)
2. Should we offer to user access to the handler allowing him to push back Actions inside the application ? (Like for the JsEvents).

cc @Zaid-Ajaj

**Comment by mastoj** [comment]
I think it looks good. 🥇 🍰

**Comment by MangelMaxime** [comment]
@codec-abc Hook support is available in fable-arch v0.10.1

**Comment by Zaid-Ajaj** [comment]
Awesome 😄
